tfhe/src/integer/gpu/server_key/radix/oprf.rs

```diff
@@ -0,0 +1,721 @@
+    /// ```
+    pub fn par_generate_oblivious_pseudo_random_signed_integer_bounded(
+        &self,
+        seed: Seed,
+        random_bits_count: u64,
+        num_blocks: u64,
+        streams: &CudaStreams,
+    ) -> CudaSignedRadixCiphertext {
+        assert!(self.message_modulus.0.is_power_of_two());
+        let range_log_size = self.message_modulus.0.ilog2() as u64 * num_blocks;
+
+        #[allow(clippy::int_plus_one)]
+        {
+            assert!(
+                random_bits_count + 1 <= range_log_size,
+                "The range asked for a random value (=[0, 2^{}[) does not fit in the available range [-2^{}, 2^{}[",
+                random_bits_count, range_log_size-1, range_log_size-1,
+            );
+        }
+
+        assert!(self.message_modulus.0.is_power_of_two());
+        let message_bits_count = self.message_modulus.0.ilog2() as u64;
+
+        let mut deterministic_seeder = DeterministicSeeder::<DefaultRandomGenerator>::new(seed);
+
+        let seeds = (0..num_blocks).map(|_| deterministic_seeder.seed());
+
+        let blocks = seeds
+            .into_iter()
+            .enumerate()
+            .map(|(i, seed)| {
+                let i = i as u64;
+                if i * message_bits_count < random_bits_count {
+                    // if we generate 5 bits of noise in n blocks of 2 bits, the third (i=2)
+                    // block must have only one bit of random
+                    if random_bits_count < (i + 1) * message_bits_count {
+                        let top_message_bits_count = random_bits_count - i * message_bits_count;
+
+                        assert!(top_message_bits_count <= message_bits_count);
+                        let ct: CudaUnsignedRadixCiphertext = self
+                            .generate_oblivious_pseudo_random(
+                                seed,
+                                top_message_bits_count,
+                                streams,
+                            );
+                        ct.ciphertext
+                    } else {
+                        let ct: CudaUnsignedRadixCiphertext = self
+                            .generate_oblivious_pseudo_random(seed, message_bits_count, streams);
+                        ct.ciphertext
+                    }
+                } else {
+                    let ct: CudaUnsignedRadixCiphertext =
+                        self.create_trivial_zero_radix(1, streams);
+                    ct.ciphertext
+                }
+            })
+            .collect::<Vec<_>>();
+
+        self.convert_radixes_vec_to_single_radix_ciphertext(&blocks, streams)
+    }
+    pub(crate) fn create_random_from_seed(
+        &self,
+        seed: Seed,
+        lwe_size: LweSize,
+    ) -> LweCiphertext<Vec<u64>> {
+        let mut ct = LweCiphertext::new(0, lwe_size, self.ciphertext_modulus);
+
+        sha3_hash(ct.get_mut_mask().as_mut(), seed);
+
+        ct
+    }
+    pub(crate) fn create_random_from_seed_modulus_switched(
```

Comment:
Doesn't this function exist in the CPU sources already? We could maybe reuse it?